main/src/library/Graph.flix

```diff
@@ -101,4 +101,68 @@ namespace Graph {
         Reachable(n1, n2) :- Reachable(n1, m), Edge(m, n2).
     }
 
+    ///
+    /// Optionally returns the shortest distance from `src` to `dst` in the weighted directed graph `g`.
+    ///
+    pub def distance(src: {src :: t}, dst: {dst :: t}, g: m[(t, d, t)]): Option[d] with Foldable[m], Boxable[t], Boxable[d], Add[d] =
+        let edges = inject g into Edge;
+        let x = query edges, dists(src.src) select x from Distance(n, x) where n == dst.dst;
```

Comment:
Instead of finding all differences and then the minimum after the fact, only the minimum distance should be saved in the first place - this can be done with lattices. The lattice instances for `Int32`, however, find the maximum instead of the minimum so:
* Either programmers with `Int32` weights should transform all weights to -w before calling distance and then `-1 * result` at the end, which is a bit awkward. For the general type `d`, the programmer would implement lattice instances opposite of intuition (min should be max, etc.)
* We wrap `d` in a wrapper enum that inverts the lattice operations, which probably makes the most sense since we want the engine to go down the lattice and find the minimum instead of the maximum.

Maybe you want to pick between these options @magnus-madsen?

In either way, the solution will use `Distance(n; d1 + d2) :- ...` in the `dists` function (maybe `Distance(n; Wrapper.+(d1, d2)) :- ...`).

Reply:
There is a lemma that shows that to compute the min distance from src to dst, you, in the general case have to compute all min distances (of reachable nodes). Thus I think this function should just call the other and then do a lookup for src and dst.